Alerting test cluster driver data race

<details>
<pre><code>
5:53AM INFO  actions.test._exec | #25 795.7 WARNING: DATA RACE

5:53AM INFO  actions.test._exec | #25 795.7 Write at 0x00c001b49c28 by goroutine 5590:

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/pkg/test.(*TestEnvAlertingClusterDriver).UninstallCluster()

5:53AM INFO  actions.test._exec | #25 795.7       /src/pkg/test/test_cluster_driver.go:345 +0xf7

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting/ops.(*AlertingOpsNode).UninstallCluster()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/ops/ops.go:129 +0x1e4

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/apis/alertops._AlertingAdmin_UninstallCluster_Handler.func1()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/apis/alertops/alertops_grpc.pb.go:217 +0x88

5:53AM INFO  actions.test._exec | #25 795.7   go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.38.0/interceptor.go:342 +0x6fa

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/apis/alertops._AlertingAdmin_UninstallCluster_Handler()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/apis/alertops/alertops_grpc.pb.go:219 +0x1dd

5:53AM INFO  actions.test._exec | #25 795.7   google.golang.org/grpc.(*Server).processUnaryRPC()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/google.golang.org/grpc@v1.52.1/server.go:1336 +0x177a

5:53AM INFO  actions.test._exec | #25 795.7   google.golang.org/grpc.(*Server).handleStream()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/google.golang.org/grpc@v1.52.1/server.go:1704 +0xff8

5:53AM INFO  actions.test._exec | #25 795.7   google.golang.org/grpc.(*Server).serveStreams.func1.2()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/google.golang.org/grpc@v1.52.1/server.go:965 +0xec

5:53AM INFO  actions.test._exec | #25 795.7

5:53AM INFO  actions.test._exec | #25 795.7 Previous read at 0x00c001b49c28 by goroutine 5586:

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/pkg/test.(*TestEnvAlertingClusterDriver).GetClusterStatus()

5:53AM INFO  actions.test._exec | #25 795.7       /src/pkg/test/test_cluster_driver.go:292 +0x8f

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting/ops.(*AlertingOpsNode).GetClusterStatus()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/ops/ops.go:111 +0x18d

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting.(*Plugin).DeleteAlertCondition()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/api_conditions.go:162 +0x275

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting.(*Plugin).onDeleteClusterAgentDisconnectHook.func1()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/sync.go:112 +0x154

5:53AM INFO  actions.test._exec | #25 795.7

5:53AM INFO  actions.test._exec | #25 795.7 Goroutine 5590 (running) created at:

5:53AM INFO  actions.test._exec | #25 795.7   google.golang.org/grpc.(*Server).serveStreams.func1()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/google.golang.org/grpc@v1.52.1/server.go:963 +0x4dd

5:53AM INFO  actions.test._exec | #25 795.7   google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/google.golang.org/grpc@v1.52.1/internal/transport/http2_server.go:598 +0x449e

5:53AM INFO  actions.test._exec | #25 795.7   google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/google.golang.org/grpc@v1.52.1/internal/transport/http2_server.go:640 +0x264

5:53AM INFO  actions.test._exec | #25 795.7   google.golang.org/grpc.(*Server).serveStreams()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/google.golang.org/grpc@v1.52.1/server.go:949 +0x2d5

5:53AM INFO  actions.test._exec | #25 795.7   google.golang.org/grpc.(*Server).handleRawConn.func1()

5:53AM INFO  actions.test._exec | #25 795.7       /go/pkg/mod/google.golang.org/grpc@v1.52.1/server.go:891 +0x64

5:53AM INFO  actions.test._exec | #25 795.7

5:53AM INFO  actions.test._exec | #25 795.7 Goroutine 5586 (running) created at:

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting.(*Plugin).onDeleteClusterAgentDisconnectHook()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/sync.go:110 +0x645

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting.(*Plugin).newClusterWatcherHooks.func4()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/impl.go:78 +0x89

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting.(*ManagementWatcherHooks[...]).HandleEvent()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/impl.go:38 +0x138

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting.(*Plugin).watchGlobalCluster()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/management.go:193 +0x1b8

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting.(*Plugin).UseWatchers.func1()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/system.go:93 +0x5d

5:53AM INFO  actions.test._exec | #25 795.7   github.com/rancher/opni/plugins/alerting/pkg/alerting.(*SimpleInternalConditionWatcher).WatchEvents.func1()

5:53AM INFO  actions.test._exec | #25 795.7       /src/plugins/alerting/pkg/alerting/impl.go:111 +0x30

5:53AM INFO  actions.test._exec | #25 795.7 ==================

5:53AM INFO  actions.test._exec | #25 795.7 • SUCCESS! 5m5.075391085s --- FAIL: TestAlerting (305.08s)

5:53AM INFO  actions.test._exec | #25 795.7     testing.go:1446: race detected during execution of test

5:53AM INFO  actions.test._exec | #25 795.7 FAIL

5:53AM INFO  actions.test._exec | #25 795.8 	github.com/rancher/opni/pkg/agent	coverage: 100.0% of statements in github.com/rancher/opni/pkg/agent, github.com/rancher/opni/pkg/gateway

5:53AM INFO  actions.test._exec | #25 795.8 	github.com/rancher/opni/pkg/gateway	coverage: 67.2% of statements in github.com/rancher/opni/pkg/agent, github.com/rancher/opni/pkg/gateway

5:53AM INFO  actions.test._exec | #25 796.4 composite coverage: 73.4% of statements

5:53AM INFO  actions.test._exec | #25 796.4

5:53AM INFO  actions.test._exec | #25 796.4 Ginkgo ran 5 suites in 6m50.022482836s

5:53AM INFO  actions.test._exec | #25 796.4

5:53AM INFO  actions.test._exec | #25 796.4 There were failures detected in the following suites:

5:53AM INFO  actions.test._exec | #25 796.4   alerting ./test/integration/alerting

5:53AM INFO  actions.test._exec | #25 796.4

5:53AM INFO  actions.test._exec | #25 796.4 Test Suite Failed

5:53AM INFO  actions.test._exec | #25 796.4 exit status 1
</pre></code>
</details>

--- a/pkg/test/test_cluster_driver.go
+++ b/pkg/test/test_cluster_driver.go
@@ -193,6 +193,7 @@ type TestEnvAlertingClusterDriver struct {
 	managedInstances []AlertingServerUnit
 	enabled          *atomic.Bool
 	ConfigFile       string
+	stateMu          *sync.RWMutex
 
 	*shared.AlertingClusterOptions
 
@@ -239,6 +240,7 @@ func NewTestEnvAlertingClusterDriver(env *Environment, opts ...alerting_drivers.
 			ResourceLimits: &alertops.ResourceLimitSpec{},
 		},
 		AlertingManagerDriverOptions: options,
+		stateMu:                      &sync.RWMutex{},
 	}
 }
 
@@ -247,6 +249,8 @@ func (l *TestEnvAlertingClusterDriver) GetClusterConfiguration(ctx context.Conte
 }
 
 func (l *TestEnvAlertingClusterDriver) ConfigureCluster(ctx context.Context, configuration *alertops.ClusterConfiguration) (*emptypb.Empty, error) {
+	l.stateMu.Lock()
+	defer l.stateMu.Unlock()
 	if err := configuration.Validate(); err != nil {
 		return nil, err
 	}
@@ -289,6 +293,8 @@ func (l *TestEnvAlertingClusterDriver) GetClusterStatus(ctx context.Context, emp
 			State: alertops.InstallState_NotInstalled,
 		}, nil
 	}
+	l.stateMu.RLock()
+	defer l.stateMu.RUnlock()
 	for _, replica := range l.managedInstances {
 		apiNode := backend.NewAlertManagerReadyClient(ctx, fmt.Sprintf("127.0.0.1:%d", replica.AlertManagerPort))
 		if err := apiNode.DoRequest(); err != nil {
@@ -310,6 +316,8 @@ func (l *TestEnvAlertingClusterDriver) InstallCluster(ctx context.Context, empty
 	if len(l.managedInstances) > 0 {
 		panic("should not have existing replicas")
 	}
+	l.stateMu.Lock()
+	defer l.stateMu.Unlock()
 	l.NumReplicas = 1
 	for i := 0; i < int(l.NumReplicas); i++ {
 		l.managedInstances = append(
@@ -339,6 +347,8 @@ func (l *TestEnvAlertingClusterDriver) InstallCluster(ctx context.Context, empty
 }
 
 func (l *TestEnvAlertingClusterDriver) UninstallCluster(ctx context.Context, req *alertops.UninstallRequest) (*emptypb.Empty, error) {
+	l.stateMu.Lock()
+	defer l.stateMu.Unlock()
 	for _, replica := range l.managedInstances {
 		replica.CancelFunc()
 	}

--- a/pkg/util/http_test.go
+++ b/pkg/util/http_test.go
@@ -111,16 +111,18 @@ func BuildHttpTransportCaching(
 			cachingHttpServer := testgrpc.NewCachingHttpServer(
 				serverPort,
 			)
-
+			shutdownchan := make(chan error, 1)
 			go func() {
 				err := cachingHttpServer.ListenAndServe()
 				if !errors.Is(err, http.ErrServerClosed) {
-					panic(err)
+					shutdownchan <- err
 				}
+				shutdownchan <- nil
 			}()
 
 			DeferCleanup(func() {
 				cachingHttpServer.Shutdown(context.TODO())
+				Expect(shutdownchan).Should(Receive(BeNil()))
 			})
 			defaultClient = http.DefaultClient
 			client := http.DefaultClient

